### PR TITLE
WIP Generating a gNMI client from the models

### PR DESCRIPTION
**Pull request by teone**
@SeanCondon just started playing with it.
I believe the same approach as for `openapi-gen` is the one makes the most sense (as you suggested yesterday).

I think I figured out how to loop through `api.Schema()` so the generation shouldn't be too hard (ready to be proven wrong tomorrow :) ), what I think is important is to agree on what the output should look like.

I scratched out a simple client in `models/testdevice-1.0.x/models/testdevice-1.0.x` (that would be `gnmiClient`) and an example usage in `main`.

I think this usage should be simple enough that it will be easily usable when imported in 3rd party applications, what do you think?

cc: @kmarquardsen

### CODE REVIEW

File: models/testdevice-1.0.x/gnmi/gnmi-gen.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+			eps, err := generateGnmiEndpointsForItem(item, itemPath)
+			if err != nil {
+				return nil, err
+			}
+			g.Endpoints = append(g.Endpoints, eps...)
+		} else if item.Kind == yang.ChoiceEntry {
+			// FIXME add support
+			log.Debug("item-is-choice-entry")
+		} else if item.IsContainer() {
+			log.Debug("item-is-container")
```
Comment (SeanCondon) [inline models/testdevice-1.0.x/gnmi/gnmi-gen.go:85]: @teone the approach looks good here

File: models/testdevice-1.0.x/gnmi/gnmi-gen.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+			_g, err := buildGnmiStruct(item, itemPath)
+			if err != nil {
+				return nil, err
+			}
+			g.Endpoints = append(g.Endpoints, _g.Endpoints...)
+		} else if item.IsList() {
+			// FIXME add support
+			log.Debug("item-is-list")
```
Comment (SeanCondon) [inline models/testdevice-1.0.x/gnmi/gnmi-gen.go:93]: the list will be the tricky bit - and the container in list, and the list in list etc

File: models/testdevice-1.0.x/scratch/gnmi-client-exploration.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+			},
+			Target: "target-foo",
+		},
+	}
+
+	getRequest := &gnmi.GetRequest{
+		Prefix:    nil,
+		Path:      leafAtTopLevelPath,
+		Type:      0,
+		Encoding:  gnmi.Encoding_PROTO,
```
Comment (SeanCondon) [inline models/testdevice-1.0.x/scratch/gnmi-client-exploration.go:102]: here you're getting a PROTO payload, but you could ask for JSON instead. The benefit of JSON is you can unmarshal it in to the YGOT object model structs. This is how it's done in the roc-api.

The tricky part comes when the get is for say `/cont1a` - the JSON returned should just be the children of this, and so you will have to unmarshal in to the appropriate YGOT struct at the right level

File: models/testdevice-1.0.x/scratch/gnmi-client-exploration.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+					Name: "leafAtTopLevel",
+				},
+			},
+			Target: "target-foo",
+		},
+	}
+
+	setRequest := &gnmi.SetRequest{
+		Update: []*gnmi.Update{
+			{
```
Comment (SeanCondon) [inline models/testdevice-1.0.x/scratch/gnmi-client-exploration.go:127]: also it should be possible to do a `jsonval` set, with a fragment of the tree serialized from YGOT structs

**Review by SeanCondon — COMMENTED** [review]
@teone the approach looks good - a few comments below

**Comment by teone** [comment]
@SeanCondon I have a basic client that is working for the `leaves` in the Yang model (with the caveat that is ignoring `ReadOnly` paths). This was pretty straight-forward.

I started to look into generating methods to manage `Containers` but that (as you noted) is way trickier.
Ideally I think it would be great to integrate the generated code with the `ygot` generated models, eg: `OnfTest1_Cont1A` in `api/generated.go`, but I'm not sure how those names/structs are generated. 
I'll start working on that tomorrow, let's see what I can get done.

For now if you want to take a look the generated code is in `api/gnmi_client.go` and an example usage is in `scratch/gnmi-client-exploration.go`

**Comment by teone** [comment]
@SeanCondon I updated the PR with a first implementation of `Get`s for container types (it's in [models/testdevice-1.0.x/api/gnmi_client.go](https://github.com/onosproject/config-models/pull/236/files#diff-99a4dc98ac231d7bd342793ec212ebd3246dcd725a208d9f7c28b53c8533c01c))

I also started exploring the set part for the containers (there is something in `models/testdevice-1.0.x/api/gnmi_client_set_cont.go`) but that will be more complex as we need to create one update for each of the nested items, so it requires reflection and it will have to be recursive. I am going to try to extract this function somewhere in go (and not in the template) so that we can unit-test it.

If you have time next week I can guide you through what I have done so far (if you prefer to go through the code yourself that's fine with me).

Let me know what you think

Teo

**Comment by teone** [comment]
@SeanCondon @tomikazi this patch is the initial implementation of the `gNMI-gen` tool.
It needs to be merged before we can generate the clients as the generated code imports utilities from [pkg/gnmi-client-gen/gnmi_utils/](https://github.com/onosproject/config-models/pull/236/files#diff-6bfbb3a7da57f087f690f173388180fca695dc326ef97e02001ca27c4d0d5bfd)

Once this is merged I can start generating the clients for all the models and smooth out all the issue we might find. There a few cases in which we are not supporting the YANG data types, one of them is `enums` but I'd like to start getting this merged and address the remaining issues in separate smaller patches.

File: pkg/gnmi-client-gen/template/_gnmi_container.go.tpl [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+    val, err := gnmi_utils.GetResponseUpdate(res)
+
+    if err != nil {
+    return nil, err
+    }
+
+    json := val.GetJsonVal()
+    st := Device{}
+    Unmarshal(json, &st)
```
Comment (SeanCondon) [inline pkg/gnmi-client-gen/template/_gnmi_container.go.tpl:33]: should handle error here

**Review by SeanCondon — APPROVED** [review]
Let's get this merged - I can see some issues (I think) like the leaf-list on leaf2e in the `testdevice-1.0.x` model, but let's come back to that later

**Pull request merged**